e2e/plugin-e2e/prometheus/utils.ts

```diff
@@ -0,0 +1,56 @@
+import { expect, type E2ESelectorGroups, type GrafanaPage } from '@grafana/plugin-e2e';
+
+/**
+ * On a page with a Prometheus query editor, sets the editor to code mode.
+ *
+ * @param grafanaPage - The Grafana page object containing the editor
+ * @param selectors - Grafana selectors object for finding UI elements
+ */
+export async function setEditorMode<P extends GrafanaPage>(
+  grafanaPage: P,
+  selectors: E2ESelectorGroups,
+  mode: 'code' | 'builder'
+) {
+  const editorModeToggle = await grafanaPage.getByGrafanaSelector(
+    selectors.components.DataSource.Prometheus.queryEditor.editorToggle
+  );
+  const codeModeButton = await editorModeToggle.locator(`css=input[id*="option-${mode}"]`); // the id contains a random string, so we use a partial match
+  await expect(codeModeButton).toBeVisible();
+  await codeModeButton.click();
+}
+
+/**
+ * Enters a query expression into the Monaco code editor and waits for Grafana to process it.
+ *
+ * @param grafanaPage - The Grafana page object containing the editor
+ * @param selectors - Grafana selectors object for finding UI elements
+ * @param expr - The query expression to enter
+ */
+export async function enterCodeEditorQueryExpr<P extends GrafanaPage>(
+  grafanaPage: P,
+  selectors: E2ESelectorGroups,
+  expr: string
+) {
+  const codeEditor = await grafanaPage
+    .getByGrafanaSelector(selectors.components.ReactMonacoEditor.editorLazy)
+    .locator('css=textarea');
+  await expect(codeEditor).toBeVisible();
+  await codeEditor.click();
+  await codeEditor.fill(expr);
+  await codeEditor.press('Tab');
+  const queryFieldContainer = await grafanaPage.getByGrafanaSelector(selectors.components.QueryField.container);
+
+  // Wait for the 'data-queryexpr' attribute to be set correctly,
+  // which indicates that the `onChange` handler has been called
+  // and the query expression has been updated. This is needed
+  // because we debounce inputs to the Monaco editor.
+  await expect(async () => {
+    const codeEditorExpr = await queryFieldContainer.getAttribute('data-queryexpr');
```

Comment:
I know it's painful to interact with Monaco in e2e tests. Even though this works, it's generally discouraged to test implementation details. You may be able to replace this with something like `await expect(grafanaPage
    .getByGrafanaSelector(selectors.components.ReactMonacoEditor.editorLazy)).toContainText(expr)`, but I'll leave it up to you to chose what's best.

Reply:
@sunker I totally agree about avoiding implementation details, and it pained me to set it up this way. Unfortunately, `await expect(grafanaPage .getByGrafanaSelector(selectors.components.ReactMonacoEditor.editorLazy)).toContainText(expr)` will return `true` before the query expression has actually updated in state, which leads to queries running with empty expressions. So I believe this is a necessary evil, but I welcome alternatives that allow us to circumvent the issue with debounced issues (documented in the code comments above).

Reply:
Ah I see. Possibly you can fill the code editor and then wait for the `Run queries` button to become enabled again. I'm not sure. Your solution is fine!